nipoppy_new/nipoppy/tabular/base.py

```diff
@@ -0,0 +1,260 @@
+                raise ValueError(
+                    f"The columns {cols} are not present in the dataframe:\n{df}"
+                )
+
+        index_self = pd.Index(zip(*[self.loc[:, col] for col in cols]))
+        index_other = pd.Index(zip(*[other.loc[:, col] for col in cols]))
+
+        diff = self.loc[~index_self.isin(index_other)]
+
+        return diff
+
+    def add_or_update_records(self, records: list[dict] | dict, validate=True) -> Self:
+        """Add or update records."""
+        if isinstance(records, dict):
+            records = [records]
+
+        # set the index (temporary)
+        self.set_index(self.index_cols, inplace=True)
+
+        # identify non-index columns
+        non_index_cols = set(self.columns) - set(self.index_cols)
+
+        for record in records:
+            # process record data
+            if validate:
+                record = self.model(**record).model_dump()
+
+            # add/update
+            for col in non_index_cols:
+                self.loc[tuple(record[col] for col in self.index_cols), col] = record[
+                    col
+                ]
+
+        self.reset_index(inplace=True)
+        return self
+
+    def concatenate(self, other: Self, validate=True) -> Self:
+        """Concatenate two dataframes."""
+        concatenated: Self = pd.concat([self, other], ignore_index=True)
+        if validate:
+            concatenated = concatenated.validate()
+        return concatenated
+
+    def save_with_backup(
+        self,
+        fpath_symlink: str | Path,
+        dname_backups: Optional[str] = None,
+        use_relative_path=True,
+        sort=True,
+    ) -> Path | None:
+        """Save the dataframe to a file with a backup."""
+        tabular_new = self.sort_values() if sort else self
+        if fpath_symlink.exists():
+            with contextlib.suppress(Exception):
+                tabular_old = self.load(fpath_symlink)
+                if sort:
+                    tabular_old = tabular_old.sort_values()
+                if tabular_new.equals(tabular_old):
+                    return None
+        return save_df_with_backup(
+            tabular_new,
+            fpath_symlink,
+            dname_backups,
+            use_relative_path,
+        )
+
+    def equals(self, other: object) -> Self:
+        """Check if two dataframes are equal."""
+        try:
+            pd.testing.assert_frame_equal(
+                self,
+                other,
+                check_like=True,
+                obj=str(self.__class__.__name__),
+            )
+            return True
+        except AssertionError:
+            return False
+
+    def sort_values(self, **kwargs):
+        """Sort the dataframe, by default on specific columns and ignoring the index."""
+        sort_kwargs = {"by": self.index_cols, "ignore_index": True}
+        sort_kwargs.update(kwargs)
+        return super().sort_values(**sort_kwargs)
+
+    def get_series_class(self) -> type[pd.Series]:
```

Comment:
Ok so this is a bit complicated but required for proper transformations/comparisons of these classes that inherit `pd.DataFrame` and `pd.Series`. It's explained in the [`pandas` docs](https://pandas.pydata.org/pandas-docs/stable/development/extending.html#extending-subclassing-pandas) which I linked in the class docstring, but in summary:
- We need the "dataframe class" (here `BaseTabular`) to contain information about 1) itself and 2) its "series class"
- This is because `pandas` has many methods that create a new dataframe from an existing one (unless we specify `inplace=True`), for example `pd.DataFrame.dropna`. If I have a `BaseTabular` dataframe then I want the result of `BaseTabular.dropna` to still be a `BaseTabular` so that it can have the extra things that I have added to the base `pd.DataFrame`. This is why I am overriding `pd.DataFrame._constructor`
- Similarly for the "series class": it needs to contain information about itself and its "dataframe class" so that e.g. if I take a slice from `BaseTabular`, then expand it into a 2D dataframe then it is still a `BaseTabular` (not just a `pd.DataFrame`. So this is what `_constructor_sliced` is for (I am overriding the `pandas` default)
- But then there is the complication of `BaseTabular` being an abstract class that other classes (e.g. `Manifest`) inherit, and each inherited class should have its own "series class" that points to the correct "dataframe class"
  - I could have just manually added the `_constructor_sliced` method to each of `Manifest`, `Bagel`, `Doughnut`, but I found a way to generate it automatically, hence the `get_series_class` method. The dictionary shenanigans with `_series_classes` are so that for example every `Manifest` has exactly the same "series class" (i.e. same ID in Python), because otherwise there are comparison issues (two series with the same data were not considered equal because the classes were technically different even though they had the same name)